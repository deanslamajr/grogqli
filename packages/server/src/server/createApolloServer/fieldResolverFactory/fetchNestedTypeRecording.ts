import {
  TypeNameToIdMapping,
  getTypeIdFromTypeName,
  getTypeRecording,
} from '../../files/type';

interface FetchNestedTypeRecordingParams {
  recordingId: string;
  typeName: string;
  typeNameToIdMappingData: TypeNameToIdMapping;
}

// TODO handle args
// TODO optimization: should only need to do this once for all root level fields of the given query execution
//  * wait for semaphore access (only allows a single access at any given time) https://www.npmjs.com/package/await-semaphore
//  * after acquiring semaphore, check cache for resolved value
//  * If exists, release semaphore and return value
//  * Else, do the work below, set cache, release semaphore, and return value
export const fetchNestedTypeRecording = async ({
  recordingId,
  typeName,
  typeNameToIdMappingData,
}: FetchNestedTypeRecordingParams) => {
  const typeId = await getTypeIdFromTypeName({
    typeName,
    typeNameToIdMappingData,
  });
  // TODO handle the case where the given typeName does not exist in the given schemaId's types.json
  if (typeId === null) {
    throw new Error(`
      TODO handle the case where the given typeName does not exist in the given types.json.
      typeName:${typeName}
    `);
  }

  const typeRecording = await getTypeRecording({
    args: {}, // TODO: replace hardcoded args value (this was done to make compiler pass)
    typeId,
    recordingId,
  });
  return typeRecording; //.value;
};
