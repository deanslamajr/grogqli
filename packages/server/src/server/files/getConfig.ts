import path from 'path';
import findUp from 'find-up';
import editJsonFile from 'edit-json-file';

const CONFIG_FILE_NAME = 'grogqli.json';

export interface GrogqliConfig {
  recordingsSaveDirectory: string;
  port: number;
  shouldDogFood: boolean;
}
const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: path.join(process.cwd(), 'grogqli'),
  port: 4000,
  shouldDogFood: false,
};

type GetGrogqliConfigValue = (configKey: keyof GrogqliConfig) => string;
let getGrogqliConfigValue: GetGrogqliConfigValue | undefined;

export const getConfig = async (): Promise<GetGrogqliConfigValue> => {
  // return early if the response is cached
  if (getGrogqliConfigValue) {
    return getGrogqliConfigValue;
  }

  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(
    async (directory) => {
      const hasConfig = await findUp.exists(
        path.join(directory, CONFIG_FILE_NAME)
      );
      return hasConfig ? directory : undefined;
    },
    { type: 'directory' }
  );

  let configFromFile: editJsonFile.JsonEditor;

  if (configLocation) {
    const configFileAbsolutePath = path.join(configLocation, CONFIG_FILE_NAME);
    configFromFile = editJsonFile(configFileAbsolutePath);
  }

  getGrogqliConfigValue = (configKey) => {
    if (configKey === undefined) {
      throw new Error(
        'Invalid use of getGrogqliConfigValue: must provide a key'
      );
    }
    let configValue: any;

    // First, try to resolve the value from the given config
    if (configLocation && configFromFile) {
      configValue = configFromFile.get(configKey);

      // Make 'recordingsSaveDirectory' an abs path
      if (
        configKey === 'recordingsSaveDirectory' &&
        configValue !== undefined
      ) {
        configValue = path.join(configLocation, configValue);
      }
    }

    // If config doesn't have a value, use the default
    if (configValue === undefined) {
      configValue = DEFAULT_CONFIG[configKey];
    }

    return configValue;
  };

  return getGrogqliConfigValue;
};
