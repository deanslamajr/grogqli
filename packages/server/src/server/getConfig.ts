import path from 'path';
import findUp from 'find-up';

const CONFIG_FILE_NAME = 'grogqli.json';

interface GrogqliConfig {
  recordingsSaveDirectory: string;
  recordingsFilename: string;
}

const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: '/home/dhan/codes/grogqli',
  recordingsFilename: 'groql_rec'
};
export const getConfig = async (): Promise<GrogqliConfig> => {
  console.log('process.cwd()', process.cwd());
  // TODO
  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(CONFIG_FILE_NAME);

  console.log('configLocation', configLocation)

  // TODO
  // if a config doesn't have a value defined for a particular lookup
  // use the value found in DEFAULT_CONFIG
  return DEFAULT_CONFIG;
}