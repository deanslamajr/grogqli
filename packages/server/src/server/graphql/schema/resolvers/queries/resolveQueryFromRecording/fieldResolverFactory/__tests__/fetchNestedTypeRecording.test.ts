import path from 'path';

import { fetchNestedTypeRecording } from '../fetchNestedTypeRecording';
import { getConfig } from '../../../../../../../getConfig';
import typeNameToIdMappingData from '../../../../__tests__/grogqli/schemas/someSchemaId/types.json';

jest.mock('../../../../../../../getConfig');

describe('fetchNestedTypeRecording', () => {
  const schemaId = 'someSchemaId';
  const recordingId = 'someTypeRecordingId';
  const typeName = 'someTypeName';

  beforeEach(() => {
    const mockedGetConfig = getConfig as jest.MockedFunction<typeof getConfig>;
    mockedGetConfig.mockImplementation(async () => () => {
      const relativePathToTestGrogqli = path.normalize(
        '../../../../__tests__/grogqli'
      );
      return path.join(__dirname, relativePathToTestGrogqli);
    });
  });

  it('should return a type recording value', async () => {
    const actual = await fetchNestedTypeRecording({
      recordingId,
      typeName,
      typeNameToIdMappingData,
    });
    expect(actual).toMatchSnapshot();
  });

  describe('if there isnt a reference to the given typeName in the given typeNameToIdMappingData', () => {
    it('should throw', async () => {
      await expect(
        fetchNestedTypeRecording({
          recordingId,
          typeName: 'nonExistentTypeName',
          typeNameToIdMappingData,
        })
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('if there isnt a type file for the given typeName', () => {
    it('should throw', async () => {
      await expect(
        fetchNestedTypeRecording({
          recordingId,
          typeName: 'doesntHaveATypeFile',
          typeNameToIdMappingData,
        })
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('if the type file for the given typeName does not have a reference to the given recordingId', () => {
    it('should throw', async () => {
      await expect(
        fetchNestedTypeRecording({
          recordingId: 'nonExistentTypeRecordingId',
          typeName,
          typeNameToIdMappingData,
        })
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
