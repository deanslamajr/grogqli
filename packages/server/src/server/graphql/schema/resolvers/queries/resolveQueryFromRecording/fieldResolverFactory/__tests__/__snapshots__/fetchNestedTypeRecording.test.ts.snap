// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchNestedTypeRecording if the type file for the given typeName does not have a reference to the given recordingId should throw 1`] = `"TODO Handle case where type file does not have a recording for the given id. typeId:someTypeId recordingId:nonExistentTypeRecordingId"`;

exports[`fetchNestedTypeRecording if there isnt a reference to the given typeName in the given typeNameToIdMappingData should throw 1`] = `
"
      TODO handle the case where the given typeName does not exist in the given types.json.
      typeName:nonExistentTypeName
    "
`;

exports[`fetchNestedTypeRecording if there isnt a type file for the given typeName should throw 1`] = `"TODO Handle case where type file does not exist. typeId:doesntHaveATypeFileId"`;

exports[`fetchNestedTypeRecording should return a type recording value 1`] = `
Object {
  "getAccountFromSession": "someAccountRecordingId",
  "getChores": "someGetChoresResponseRecordingId",
}
`;
